### ISSUE TITLE
Hash collision between labels will lead to incorrect results

### ISSUE THREAD

**Issue by wzharies** [prometheus/prometheus#12519]
### What did you do?

After reading the code related to stringlabel, I found that there are a lot of hash function calls, and it seems that the errors caused by hash collisions are not considered.
For example in funcLabelReplace:
https://github.com/prometheus/prometheus/blob/031d22df9e43f1c3014b8344337e133099902ae8/promql/functions.go#L1149-L1201
And there are many ContainsSameLabelset checks in the calculation of eval, here is also judged by hash, will there be the same problem?
https://github.com/prometheus/prometheus/blob/031d22df9e43f1c3014b8344337e133099902ae8/promql/value.go#L232-L249

### What did you see instead? Under which circumstances?

I'm not sure if it's really a problem, but I really don't see how collisions are handled

**Comment by bboreham** [comment]
You're correct: there are a number of places in Prometheus where Labels hash collisions are not handled.

They seem to be quite rare in the wild; #5724 is the only example I can see reported here. By ["birthday problem" math](https://en.wikipedia.org/wiki/Birthday_problem) with 6 million series there is a 1 in a million chance of a collision.

Working on #12993 gave me the idea that we could produce a 256-bit hash for every series, then use that everywhere that the 64-bit hash is currently used. Collisions would be astronomically unlikely. 

A (likely incomplete) list of places `Labels.Hash()` is used without any protection against collisions:
* ~TSDB, to find a series in the head block.~ - this one does handle collisions.
* In the PromQL engine, collecting the results of a range query.
* `ContainsSameLabelset`, to check output of a PromQL subexpression.
* `DropMetricName`, used by most PromQL functions.
* As the key to `scrapePool.loops` and `scrapePool.activeTargets` (target labels and URL are hashed together).

There are also hashes of subsets of labels, which would not be helped by pre-computing a 256-bit hash.
* Grouping in PromQL aggregations like `sum by (foo, bar)`

**Comment by beorn7** [comment]
I must admit that the untreated hash collisions are still giving me the creeps.

My main concern is that we often use fast non-cryptographic hashes, which are probably much farther away from perfect pseudo-random distribution of hash values than cryptographic hashes. Pardon me if I use the wrong terminology here (I'm not a proper statistician), but the gist is that the probability of a collision is much higher than calculated via the birthday problem math if the hash values are correlated with the original values rather than randomly and equally distributed.

In prometheus/prometheus, we use xxHash, which is probably better than fnv64a, which we use in prometheus/client_golang. (See https://github.com/prometheus/client_golang/issues/220 for the hash collision handling added in client_golang eventually, but I'm not so sure anymore it deals with all cases properly.)

While properly dealing with all variations of hash collisions in a performance-preserving way would be cool, a sufficiently low chance of hash collisions is also "good enough". Note that the only real-world example reported by @bboreham above was fixed by improving the hash calculation (when combining different hash values).

**Comment by bwplotka** [comment]
Hello from the Bug Scrub!

Sounds like it's still relevant, but not easy. I like the idea of increasing bit size of a hash (or using more secure one) vs handling collisions, but proper proposal and experiment with benchmark is needed.
